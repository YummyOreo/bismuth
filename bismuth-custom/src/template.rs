#[derive(Default, Debug)]
pub struct Template {
    pub content: String,
}

impl Template {
    pub fn new(content: String) -> Self {
        Template { content }
    }
}
