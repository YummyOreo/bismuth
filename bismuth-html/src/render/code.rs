use syntect::easy::HighlightLines;
use syntect::highlighting::{Color, ThemeSet};
use syntect::html::highlighted_html_for_string;
use syntect::parsing::SyntaxSet;

use syntect::Error;
use thiserror::Error;

fn init() -> (SyntaxSet, ThemeSet) {
    let ps = SyntaxSet::load_defaults_newlines();
    let ts = ThemeSet::load_defaults();

    (ps, ts)
}

// #[derive(Error, Debug)]
// #[error("Could not find lang: `{lang}`")]
// pub struct FindLangError {
//     lang: String,
// }

#[derive(Debug, Error)]
pub enum HighlightError {
    #[error("Could not find lang: `{lang}`")]
    FindLangError { lang: String },
    #[error("Internal error {0}")]
    Internal(#[from] Error),
}

pub fn highlight(lang: String, code: String) -> Result<String, HighlightError> {
    let (ps, ts) = init();
    let syntax = match ps.find_syntax_by_extension(&lang) {
        Some(syntax) => syntax,
        None => ps
            .find_syntax_by_name(&lang)
            .ok_or(HighlightError::FindLangError { lang })?,
    };

    highlighted_html_for_string(&code, &ps, syntax, &ts.themes["InspiredGitHub"])
        .map_err(HighlightError::Internal)
}

#[cfg(test)]
mod test {

    use super::*;
    macro_rules! snapshot {
        ($content:tt) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../../testdata/output/render/code/");
            settings.bind(|| {
                insta::assert_snapshot!($content);
            });
        };
    }

    #[test]
    fn test() {
        let code = "fn test() {\n\tlet a = 10;\n}".to_string();
        let html = highlight("rs".to_string(), code).unwrap();
        snapshot!(html);
    }

    #[test]
    fn test_inline() {
        let code = "let test = 10".to_string();
        let html = highlight("rs".to_string(), code).unwrap();
        snapshot!(html);
    }
}
