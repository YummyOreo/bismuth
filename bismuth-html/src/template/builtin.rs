pub const TEST: &str = "Test template: {value_1} {value_2} {elements}";
pub const DEFAULT: &str = "{elements}";
