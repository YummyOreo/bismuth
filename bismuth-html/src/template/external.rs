//! Stuff for external templates
//! This also includes default templates
