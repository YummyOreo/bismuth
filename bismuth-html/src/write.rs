#![allow(dead_code)]
// use crate::render::Renderer;

const BUILD: &str = "./build/";
const BUILD_ASSETS: &str = "./build/assets";
const ASSETS: &str = "./assets";

pub mod utils {
    use super::*;
    use std::{
        fs,
        io::Error,
        path::{Path, PathBuf},
    };

    /// Makes the `./build/` folder along with the `./bulid/assets/` folder
    pub fn make_build() -> Result<(), Error> {
        let build = Path::new(BUILD);
        if !build.exists() {
            fs::create_dir_all(&build)?
        }
        make_build_assets()?;
        Ok(())
    }

    /// Makes the `./bulid/assets/` folder
    /// Will make the `./build/` folder if it does not exitst
    /// But if it does not, you *should* call `make_build()`
    pub fn make_build_assets() -> Result<(), Error> {
        let assets = Path::new(BUILD_ASSETS);
        if !assets.exists() {
            fs::create_dir_all(&assets)?
        }
        Ok(())
    }

    /// Makes the `./assets/` folder
    pub fn make_assets() -> Result<(), Error> {
        let assets = Path::new(ASSETS);
        if !assets.exists() {
            fs::create_dir_all(&assets)?
        }
        Ok(())
    }

    /// Writes a html file to the `./build/` folder
    /// The path should not have a . in the begining
    pub fn write_html_file(content: &str, path: &PathBuf) -> Result<(), Error> {
        // Makes build dir if it does not exitst
        make_build()?;

        let path_build = Path::new(BUILD);
        let full_path = path_build.join(path);

        fs::write(full_path, content)
    }

    /// Moves a asset from the `./assets/` folder to the `./bulid/assets/` folder
    /// The path should not have a . in the begining
    pub fn move_asset(path: &PathBuf) -> Result<(), Error> {
        // Makes both the `./assets/` folder and the `./build/assets/` folder if the do not exitst
        make_assets()?;
        make_build_assets()?;

        let new_path = Path::new(BUILD).join(&path);
        let old_full = Path::new("./").canonicalize()?.join(&path);

        fs::copy(old_full, new_path).map(|_| ())
    }
}
