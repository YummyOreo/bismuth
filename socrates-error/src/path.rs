#![allow(unused)]
use std::{io::stdout, path::PathBuf};

use crossterm::{
    cursor,
    event::{read, Event, KeyCode},
    execute,
    style::{Color, Print, ResetColor, SetForegroundColor},
    terminal, Result,
};

use crate::error_ui;

fn update_input(path: &mut String) -> Option<()> {
    if let Event::Key(k) = read().expect("Should be able to read") {
        match k.code {
            KeyCode::Char(x) => {
                path.push(x);
                execute!(stdout(), Print(x.to_string()));
            }
            KeyCode::Backspace => {
                path.pop();
                let print_path = format!("Enter new path: {}", path);
                execute!(
                    stdout(),
                    terminal::Clear(terminal::ClearType::CurrentLine),
                    cursor::MoveToColumn(0),
                    Print(print_path),
                );
            }
            KeyCode::Enter => return Some(()),
            _ => {}
        }
    }
    None
}

pub fn md_file_error(_description: &str) -> Option<PathBuf> {
    let options = vec![
        // "Make folder at path.".to_string(),
        "Chose another path".to_string(),
    ];

    let option = error_ui(&options)?;

    match option {
        // 1 => {}
        1 => {
            execute!(
                stdout(),
                cursor::MoveUp(u16::try_from(options.len()).unwrap() + 1_u16),
                terminal::Clear(terminal::ClearType::CurrentLine),
                cursor::SavePosition
            );
            for i in 0..options.len() + 1 {
                execute!(
                    stdout(),
                    cursor::MoveDown(1),
                    terminal::Clear(terminal::ClearType::CurrentLine),
                    cursor::RestorePosition,
                );
            }

            execute!(
                stdout(),
                ResetColor,
                Print("Enter new path: "),
                cursor::Show
            );
            let mut path = "".to_string();
            loop {
                if update_input(&mut path).is_some() {
                    break;
                }
            }
            return Some(PathBuf::from(path));
        }
        _ => panic!("Something has gone terribly wrong"),
    }
    None
}
