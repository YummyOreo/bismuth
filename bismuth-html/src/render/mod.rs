use bismuth_parser::{
    tree::{Element, Kind},
    Parser,
};
use regex::Regex;
use std::convert::TryFrom;
use std::path::{Path, PathBuf};

mod code;
use crate::render::code::highlight;
use crate::template::Template;
use crate::write::{move_assets, utils::write_html_file};

const URL_CHECK: &str =
    r"^(http(s)://.)?[-a-zA-Z0-9@:%._\+~#=]{2,256}\.[a-z]{2,6}\b([-a-zA-Z0-9@:%_\+.~#?&//=]*)$";

pub trait Render {
    fn render(&mut self, path: &Path) -> Option<String>;
}

#[derive(Clone, Debug)]
pub struct Renderer {
    pub parser: Parser,
    pub asset_list: Vec<PathBuf>,

    output: String,

    /// This is the path that the file will be placed to
    /// Ie `.../output/test/index.html`
    path: PathBuf,
}

impl Renderer {
    /// You should pass parser through bismuth_custom::parse_custom() first
    /// Then put the output into this
    pub fn new(parser: Parser) -> Self {
        let path = PathBuf::from(
            parser
                .metadata
                .frontmatter
                .get_path()
                .expect("Should have a path"),
        );
        Self {
            parser,
            asset_list: vec![],
            output: String::new(),
            path,
        }
    }

    /// Will move the assets if self.output has stuff in it, and if the asset_list is not empty
    /// Will return Ok(true) if it attempted to move it
    /// Will return Ok(false) if self.output is empty
    pub fn move_assets(&self) -> Result<bool, std::io::Error> {
        if self.output.is_empty() {
            return Ok(false);
        } else if self.asset_list.is_empty() {
            return Ok(true);
        }
        move_assets(&self.asset_list)?;
        Ok(true)
    }

    pub fn write(&self) -> Result<(), std::io::Error> {
        self.move_assets()?;
        write_html_file(
            &self.output,
            &self.path,
            &self
                .parser
                .metadata
                .frontmatter
                .get_title()
                .cloned()
                .unwrap(),
        )
    }
}

/// This will set self.output for you
/// asset_list will be populated with the assets that are needed to be moved
impl Render for Renderer {
    fn render(&mut self, _path: &Path) -> Option<String> {
        let kind = self.parser.metadata.frontmatter.get_kind()?;

        let mut values = self
            .parser
            .metadata
            .frontmatter
            .get_values()
            .unwrap_or_default();
        if let Some(title) = self.parser.metadata.frontmatter.get_title().cloned() {
            values.insert(String::from("title"), title);
        }

        let elements = &self.parser.ast.elements;
        let mut template = Template::new_from_name(kind, &values, None, elements)?;

        self.output = template.render(&self.path)?;

        // --- HORRIBLE PLS REPLACE WITH GOOD STUFF ---
        // replace all double br's + w/ <double br>
        let replace_double_br = Regex::new(r"(<br>\n*){2}").unwrap();
        self.output = replace_double_br
            .replace_all(&self.output, "<double br>\n")
            .to_string();

        // Remove all other br's
        let remove_br = Regex::new(r"<br>\n?").unwrap();
        self.output = remove_br.replace_all(&self.output, "\n").to_string();

        // Replace all double br's w/ single br's
        let replace_br = Regex::new(r"<double br>\n").unwrap();
        self.output = replace_br.replace_all(&self.output, "<br>\n").to_string();

        self.asset_list.append(&mut template.asset_list);
        Some(self.output.clone())
    }
}

/// Returns (Html, File to move)
fn handle_file_url(url: &str, text: &str, _path: &Path) -> (String, Option<PathBuf>) {
    let valid_url = Regex::new(URL_CHECK).expect("Should be valid regex");

    if valid_url.is_match(url) {
        return (format!(r#"<img src="{url}" alt="{text}">"#), None);
    } else {
        let picture_rg = Regex::new(r"^.+\.(png|jpeg|apng|avif|gif|jpg|jfif|pjpeg|pjp|svg|webp)$")
            .expect("Should be valid regex");
        let video_rg = Regex::new(r"^.+\.(webm|mp4)$").expect("Should be valid regex");

        if picture_rg.is_match(url) {
            return (
                format!(r#"<img src="{url}" alt="{text}">"#),
                Some(PathBuf::from(url)),
            );
        } else if video_rg.is_match(url) {
            let format = video_rg
                .captures_iter(url)
                .next()
                .expect("Should have 2 capture groups")
                .get(1)
                .expect("Should have 1 capture")
                .as_str();
            return (
                format!(r#"<source src="{url}" type="video/{format}">"#),
                Some(PathBuf::from(url)),
            );
        }
    }
    Default::default()
}

/// This will be relitive to the base dir
fn handle_link(url: &str, text: &str) -> (String, String) {
    let valid_url = Regex::new(URL_CHECK).expect("Should be valid regex");

    if valid_url.is_match(url) {
        (
            format!(r#"<a href="{}" target="_blank">{}"#, url, text),
            r"</a>".to_string(),
        )
    } else {
        (
            format!(r#"<a href="{}">{}"#, url, text),
            r"</a>".to_string(),
        )
    }
}

impl Render for Element {
    fn render(&mut self, path: &Path) -> Option<String> {
        let mut inside = self
            .elements
            .iter()
            .map(|e| e.clone().render(path).expect("This should not fail"))
            .collect::<String>();

        // Gets the html of the kind. Some kinds (like Text) may not have a end
        let (start, end) = match &self.kind {
            Kind::Paragraph => (String::from("<p>"), String::from("</p>")),
            Kind::Bold => (String::from("<b>"), String::from("</b>")),
            Kind::Italic => (String::from("<i>"), String::from("</i>")),
            Kind::BoldItalic => (String::from("<b><i>"), String::from("</i></b>")),
            Kind::Blockquote => (String::from("<blockquote>"), String::from("</blockquote>")),
            Kind::Header => {
                let mut num = self.get_attr("level").cloned().unwrap_or(String::from("6"));
                if num.parse::<i8>().unwrap_or_default() > 6_i8 {
                    num = String::from("6");
                }
                let rg = regex::Regex::new(r"\s+").unwrap();
                let pre_text = html_escape::encode_text(inside.trim());
                let text = rg.replace_all(&pre_text, "-").to_lowercase();
                (
                    format!(r##"<h{num} id="{text}"><a href="#{text}">"##),
                    format!("</a></h{num}>"),
                )
            }
            Kind::Text => (self.text.clone().unwrap_or_default(), Default::default()),

            Kind::Link => handle_link(
                &self.get_attr("link").cloned().unwrap_or_default(),
                &self.get_text().cloned().unwrap_or_default(),
            ),
            Kind::FilePrev => {
                let (html, asset) = handle_file_url(
                    &self.get_attr("link").cloned().unwrap_or_default(),
                    &self.text.clone().unwrap_or_default(),
                    path,
                );
                if let Some(asset) = asset {
                    self.asset_list.push(asset);
                }
                (html, Default::default())
            }

            Kind::ListItem => (
                format!(
                    r#"<li class="item">{}"#,
                    String::from("\t").repeat(
                        self.get_attr("level")
                            .cloned()
                            .unwrap_or(String::from("1"))
                            .parse::<usize>()
                            .unwrap()
                            + 1_usize
                    )
                ),
                String::from("</li>"),
            ),
            Kind::OrderedListElement => (
                format!(
                    r#"<li class="num-list">{}{}."#,
                    String::from("\t").repeat(
                        self.get_attr("level")
                            .cloned()
                            .unwrap_or(String::from("1"))
                            .parse::<usize>()
                            .unwrap()
                            + 1_usize
                    ),
                    self.get_attr("num").cloned().unwrap_or(String::from("0"))
                ),
                String::from("</li>"),
            ),

            Kind::InlineCode => (
                format!(
                    r#"<div class="inline-code">{}"#,
                    self.text.clone().unwrap_or_default()
                ),
                String::from("</div>"),
            ),
            Kind::BlockCode => (
                highlight(
                    self.get_attr("lang")
                        .cloned()
                        .unwrap_or(String::from("plaintext")),
                    self.text.clone().unwrap_or_default(),
                )
                .unwrap(),
                Default::default(),
            ),

            Kind::HorizontalRule => (String::from("<hr>"), Default::default()),
            Kind::EndOfLine => (String::from("\n<br>\n"), Default::default()),

            Kind::InlineLaTeX => (
                katex::render_with_opts(
                    &self.get_text().cloned().unwrap_or_default(),
                    katex::Opts::builder()
                        .output_type(katex::OutputType::Mathml)
                        .build()
                        .unwrap(),
                )
                .unwrap(),
                Default::default(),
            ),
            Kind::BlockLaTeX => (
                format!(
                    "<br>\n<div class=\"latex\">{}",
                    katex::render_with_opts(
                        &self.get_text().cloned().unwrap_or_default(),
                        katex::Opts::builder()
                            .output_type(katex::OutputType::Mathml)
                            .display_mode(true)
                            .build()
                            .unwrap(),
                    )
                    .unwrap()
                ),
                String::from("\n</div>\n<br>"),
            ),
            Kind::CustomElement(_c) => {
                if let Ok(mut t) = Template::try_from(&self.to_owned()) {
                    // Remove already rendered elements
                    inside = String::new();
                    let html = t.render(path)?;
                    self.asset_list.append(&mut t.asset_list);
                    (html, Default::default())
                } else {
                    Default::default()
                }
            }
            _ => Default::default(),
        };
        Some(format!("{start}{inside}{end}"))
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use std::fs;

    use bismuth_custom::parse_custom;

    fn snapshot(content: &str) -> String {
        let mut parser = Parser::new_test("/test/test.md", content);
        parser.parse().unwrap();
        let parser = parse_custom(parser, &vec![]);
        let mut render = Renderer::new(parser);
        render.render(&PathBuf::new()).unwrap()
    }

    macro_rules! snapshot {
        ($name:tt, $content:tt) => {
            #[test]
            fn $name() {
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../../testdata/output/render/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot($content));
                });
            }
        };
    }

    macro_rules! snapshot_path {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let path = PathBuf::from($path);
                println!("{path:?}");
                let content = fs::read_to_string(&path).unwrap();
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../../testdata/output/render/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot(&content));
                });
            }
        };
    }

    snapshot!(
        test,
        "test *test* \n```rust\nfn test() {\n\tprintln!(\"test\")\n}\n```"
    );

    snapshot!(
        test_2,
        "# hearder\n- 1\n    - 2\n1. list item\nthis is a *__good test__*!! \n `inline?`\n---\n> blockquote"
    );

    snapshot!(test_3, "***test?***");

    snapshot!(test_br, "test test \n\n\ntest test\ntest\n");
    snapshot!(
        test_latex,
        "this is a test for inline latex using katex: $E = mc^2$"
    );

    snapshot!(
        test_latex_1,
        "this is a test for block latex using katex:\n$$E = mc^2$$"
    );

    snapshot_path!(test_path, "./testdata/test/render/test.md");
}
