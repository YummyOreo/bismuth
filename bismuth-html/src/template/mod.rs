#![allow(unused, dead_code)]
use bismuth_parser::{
    custom::CustomElm,
    tree::{Element, Kind},
    Parser,
};
use regex::Regex;
use std::collections::HashMap;
use std::convert::TryFrom;
use std::path::PathBuf;

use crate::render::{Render, Renderer};

pub mod builtin;

/// Data/Values: Will be replaced (as a string) at {key}
/// Reserved keys: `body` and `elements` (Sometimes title if a full page template)
///
/// Body: Will be replaced at {body}
/// Elements: Will be replaced at {elements}
#[derive(Debug, PartialEq)]
pub struct Template<'a> {
    template: &'a str,
    values: &'a HashMap<String, String>,
    body: Option<&'a String>,
    pub elements: &'a Vec<Element>,
    pub asset_list: Vec<PathBuf>,
}

impl<'a> TryFrom<&'a Element> for Template<'a> {
    type Error = ();

    /// Try to convert a element that's kind == CustomElement to a Template
    /// If it is a custom element, it will also try to get the template
    /// If it can't get it, it will return a Error
    fn try_from(elm: &'a Element) -> Result<Self, Self::Error> {
        if let Kind::CustomElement(c) = &elm.kind {
            if let Some(t) = &c.template {
                return Ok(Self {
                    template: t,
                    values: &c.values,
                    body: c.body.as_ref(),
                    elements: &elm.elements,
                    asset_list: vec![],
                });
            }
        }
        Err(())
    }
}

impl<'a> Template<'a> {
    pub fn new(
        template_str: &'a str,
        values: &'a HashMap<String, String>,
        body: Option<&'a String>,
        elements: &'a Vec<Element>,
    ) -> Self {
        Self {
            template: template_str,
            values,
            body,
            elements,
            asset_list: vec![],
        }
    }

    pub fn get_template(name: &str) -> Option<&str> {
        match name.to_lowercase().as_str() {
            "test" => Some(builtin::TEST),
            "default" => Some(builtin::DEFAULT),
            _ => None,
        }
    }

    pub fn new_from_name(
        name: &'a str,
        values: &'a HashMap<String, String>,
        body: Option<&'a String>,
        elements: &'a Vec<Element>,
    ) -> Option<Self> {
        let template_str = Self::get_template(name)?;

        Some(Self::new(template_str, values, body, elements))
    }
}

impl Render for Template<'_> {
    fn render(&mut self, path: &PathBuf) -> Option<String> {
        let mut output = self.template.to_string();
        // First replace {elements} w/ rendered elements
        let mut elements_str = self
            .elements
            .iter()
            .map(|e| {
                let mut element = e.clone();
                let html = element.render(path).expect("Should not fail");
                self.asset_list.append(&mut element.asset_list);
                html
            })
            .collect::<String>();
        let e_rg = Regex::new(r"\{(?i)elements\}").expect("Should be valid regex");
        output = e_rg.replace(&output, elements_str).to_string();

        // Next do the body
        let b_rg = Regex::new(r"\{(?i)body\}").expect("Should be valid regex");
        let body_default = String::new();
        output = b_rg
            .replace(&output, self.body.cloned().unwrap_or_default())
            .to_string();

        // next do each value
        for (key, value) in self.values {
            let rg = Regex::new(&format!(r"\{{(?i){}\}}", key)).expect("Should be valid");
            output = rg.replace(&output, value).to_string();
        }
        Some(output)
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use bismuth_parser::Parser;

    fn init_parser<'a>(content: &'a str, frontmatter: &'a str) -> Parser {
        let mut parser = Parser::new_test("/test/", content);
        parser
            .metadata
            .frontmatter
            .update_from_str(frontmatter)
            .unwrap();

        parser.parse();
        parser
    }

    macro_rules! init_template {
        ($parser:tt, $template:expr, $body:expr) => {
            Template {
                template: $template,
                body: $body,
                elements: &$parser.ast.elements,
                values: &$parser.metadata.frontmatter.get_values().unwrap(),
                asset_list: vec![],
            }
        };
    }

    macro_rules! init_template_name {
        ($parser:tt, $name:expr, $body:expr) => {
            Template {
                template: Template::get_template(&$name).unwrap(),
                body: $body,
                elements: &$parser.ast.elements,
                values: &$parser.metadata.frontmatter.get_values().unwrap(),
                asset_list: vec![],
            }
        };
    }

    macro_rules! snapshot {
        ($content:tt) => {
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("../../testdata/output/template/");
            settings.bind(|| {
                insta::assert_snapshot!($content);
            });
        };
    }

    #[test]
    fn test() {
        let parser = init_parser(
            "{value} {another_value}",
            "
            values:
                - value: test
                - another_value: test another value
            ",
        );
        let mut template = init_template!(parser, &String::from("test:\n {elements}"), None);

        let s = template.render(&PathBuf::new()).unwrap();
        snapshot!(s);
    }

    #[test]
    fn test_1() {
        let parser = init_parser(
            "{value}} {hmm}",
            "
            values:
                - value}: test
                - another_value: test another value
            ",
        );
        let mut template = init_template!(parser, &String::from("test:\n {elements}"), None);

        let s = template.render(&PathBuf::new()).unwrap();
        snapshot!(s);
    }

    #[test]
    fn test_2() {
        let parser = init_parser(
            "*test*",
            "
            kind: test
            values:
                - value_1: test
                - value_2: test another value
            ",
        );
        let name = parser.metadata.frontmatter.get_kind().unwrap();
        let mut template = init_template_name!(parser, name, None);

        let s = template.render(&PathBuf::new()).unwrap();
        snapshot!(s);
    }
}
