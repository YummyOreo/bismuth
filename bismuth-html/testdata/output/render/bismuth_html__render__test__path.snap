---
source: bismuth-html/src/render/mod.rs
expression: snapshot(&content)
---
<!DOCTYPE html>
<html lang="">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width,initial-scale=1.0">
    <meta name="head:count" content="4">
    <link rel="stylesheet" href="/assets/css/colors.css">
    <link rel="stylesheet" href="/assets/css/base.css">
    <title>test</title>
</head>

<body>
    <div class="content">
        <div class="inner">
            <h1 id="h1-heading"><a href="#h1-heading"> h1 Heading</a></h1>

<h2 id="h2-heading"><a href="#h2-heading"> h2 Heading</a></h2>

<h3 id="h3-heading"><a href="#h3-heading"> h3 Heading</a></h3>

<h4 id="h4-heading"><a href="#h4-heading"> h4 Heading</a></h4>

<h5 id="h5-heading"><a href="#h5-heading"> h5 Heading</a></h5>

<h6 id="h6-heading"><a href="#h6-heading"> h6 Heading</a></h6>
<br>

<h2 id="horizontal-rules"><a href="#horizontal-rules"> Horizontal Rules</a></h2>

<blockquote> This is cool</blockquote>
<br>
<hr>
<br>
<p>!!!!!! ???? ,,  -- ---</p>
<br>
<p>"Smartypants, double quotes" and 'single quotes'</p>
<br>

<h2 id="emphasis"><a href="#emphasis"> Emphasis</a></h2>
<br>
<b>This is bold text</b>
<br>
<b>This is bold text</b>
<br>
<i>This is italic text</i>
<br>
<i>This is italic text</i>
<br>
<h2 id="blockquotes"><a href="#blockquotes"> Blockquotes</a></h2>
<br>

<blockquote> Blockquote...</blockquote>
<br>

<h2 id="lists"><a href="#lists"> Lists</a></h2>
<br>
<p>Unordered</p>
<br>
<li class="item">	 Sub-lists are made by indenting 2 spaces:</li>

<li class="item">		 Marker character change forces new list start:</li>

<li class="item">			 Facilisis in pretium nisl aliquet</li>

<li class="item">			 Nulla volutpat aliquam velit</li>

<li class="item">	 Very easy!</li>
<br>
<p>Ordered</p>
<br>
<li class="num-list">	1. Lorem ipsum dolor sit amet</li>

<li class="num-list">	2. Consectetur adipiscing elit</li>

<li class="num-list">	3. Integer molestie lorem at massa</li>
<br>

<li class="num-list">	1. You can use sequential numbers...</li>

<li class="num-list">	1. ...or keep all the numbers as <div class="inline-code">1.</div></li>
<br>
<p>Start numbering with offset:</p>
<br>
<li class="num-list">	57. foo</li>

<li class="num-list">	1. bar</li>
<br>

<h2 id="code"><a href="#code"> Code</a></h2>
<br>
<p>Inline <div class="inline-code">code</div></p>
<br>
<p>Block code "fences"</p>
<br>
<pre style="background-color:#ffffff;">
<span style="color:#323232;">
</span><span style="color:#323232;">Sample text here...
</span></pre>

<br>
<p>Syntax highlighting</p>
<br>
<pre style="background-color:#ffffff;">
<span style="color:#323232;">
</span><span style="font-weight:bold;color:#a71d5d;">var </span><span style="font-weight:bold;color:#795da3;">foo </span><span style="font-weight:bold;color:#a71d5d;">= function </span><span style="color:#323232;">(bar) {
</span><span style="color:#323232;">  </span><span style="font-weight:bold;color:#a71d5d;">return </span><span style="color:#323232;">bar</span><span style="font-weight:bold;color:#a71d5d;">++</span><span style="color:#323232;">;
</span><span style="color:#323232;">};
</span><span style="color:#323232;">
</span><span style="color:#795da3;">console</span><span style="color:#323232;">.</span><span style="color:#0086b3;">log</span><span style="color:#323232;">(foo(</span><span style="color:#0086b3;">5</span><span style="color:#323232;">));
</span></pre>

<br>
<h2 id="links"><a href="#links"> Links</a></h2>
<br>
<a href="/test">link text</a>
<br>
<a href="example.com" target="_blank">link with title</a>
<br>
<h2 id="images"><a href="#images"> Images</a></h2>
<br>
<img src="/assets/test.png" alt="Minion">

<img src="example.com" alt="Stormtroopocat">
<br>
<p>Like links, Images also have a footnote style syntax</p>
<br>
<h2 id="custom-stuff:"><a href="#custom-stuff:"> Custom stuff:</a></h2>
<br>

<h3 id="tex:"><a href="#tex:"> Tex:</a></h3>
<br>
<p>inline <span class="katex"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>T</mi><mi>e</mi><mi>x</mi></mrow><annotation encoding="application/x-tex">Tex</annotation></semantics></math></span></p>
<br>

<div class="latex"><span class="katex"><math xmlns="http://www.w3.org/1998/Math/MathML" display="block"><semantics><mrow><mi>N</mi><mi>o</mi><mi>n</mi><mi>e</mi><mi>i</mi><mi>n</mi><mi>l</mi><mi>i</mi><mi>n</mi><mi>e</mi></mrow><annotation encoding="application/x-tex">None inline</annotation></semantics></math></span>
</div>
<br>



        </div>
    </div>
</body>

</html>

